[DRAFT] THE REWRITE!

covers1624 opened the pull request:
WorkspaceTool was one of my first non-Minecraft projects. There is a LOT of baggage hanging around. It's in dire need of a mass cleanup. There's no unit testing, things are fragile, and with Gradle 8 for modding around the corner, things WILL explode.

So, let's rewrite the thing!

Goals:
- Rewrite from scratch (mostly). Some things may be similar/the same.
- ~~Preserve existing script syntax (mostly).~~
  - I have decided that a more data-driven approach to the previous script based system is needed. I have settled on Yaml for a more human friendly configuration format. Internally this is just using JSON/Gson, so I could open it up to JSON as well.
- ~~Don't drop any features.~~
  - Unfortunately, Older Forgedev workspaces may/will be dropped initially. I do still want to keep my 1.12.2 environment alive, so perhaps only intermediate versions (1.14, 1.15) etc, may be dropped in the end.
- ~~Properly extract Gradle configuration data so we can mutate it.~~
  - Perform any mutation of the Gradle state, gradle side, before extracting more simplified data.
- Prepare for multi-loader support.
- UNIT AND INTEGRATION TESTING!
  -  Test all the gradle extract stuff. With/without FG, etc.
- Keep Intellij plugin integration in mind.
  
  
TODO:
- [x] Add Gralde exporter classpath hashes to the exported Gradle data caches. So we can invalidate it.
- [x] Build an integration testing framework, this framework should ideally commit the outputted IntelliJ files to the repository for validation. Perhaps similar work to what I did in CoffeeGrinder could be used here.
- [x] Fix extraction of some source set dependencies.
  - [x] The `test` source set has the `main` source set output directly added to the `compileClasspath` file collection instead of to `testCompileClasspath` configuration. We should be able to deal with these.
  - [x] Some Neoforge versions use `implementation project(path: ':neoforge', configuration: 'runtimeElements')` these don't currently get detected.
- [ ] Extract publishing information to replace `archivesBaseName` is old and likely not sufficient anymore.
  - [ ] We _may_ be able to track which source set a published jar is produced from. I'm not sure yet.
- [ ] Extract exported `api` dependencies, and properly mark them as such for each module. This may tie in with publishing data extraction.
- [ ] Implement Run Configuration exporting.

Bugs:
- [x] We currently emit the NeoForge module content roots in such a way that multiple modules have the same content roots, this is invalid in Intellij and prevnets us from editing the project after import.
- [x] Sub-projects build.gradle files don't get included in the hash of the gradle project. Perhaps we need to have Gradle feedback the list of important cacheable files?

This is obviously still a work in progress. The above list may grow (or shrink).

covers1624 committed:
NUKE EVERYTHING. Prepare to start from scratch!

covers1624 committed:
Start the rewrite of gradle extraction data.

Officialy commented:
I love you and this project, my happiness peaked when I got the notification email :p

covers1624 committed:
Use static class for dummy WorkspaceToolModel impl.

covers1624 committed:
Fix issues with plugin id extraction.

covers1624 committed:
Fix wrong key for plugin data.

covers1624 committed:
Setup junit and annotations for all sub-projects.

covers1624 committed:
Setup dev plugin data to be loaded from the classpath instead of sysprop

covers1624 committed:
Add test-api module and add to all test runtime classpaths.

covers1624 committed:
Actually return SourceSet data from Gradle.

covers1624 committed:
Initial working impl of GradleModelExtractor, basic unit tests.

covers1624 committed:
Declare genGradlePluginMetaDev to run on Gradle import.

covers1624 committed:
Run Gradle with the correct Java version.

covers1624 committed:
Make test task depend on genGradlePluginMetaDev

covers1624 committed:
Proper names for gradle cache files.

covers1624 committed:
Update Gradle tooling api, properly stop Gradle daemons.

covers1624 committed:
Allow manually enabling Gradle daemon jvm attach from unit tests.

covers1624 committed:
Clarify SourceSetContainer deprecation/removal.

covers1624 committed:
Refactors and configuration/dependency extraction.

covers1624 committed:
Specify we want Temurin/AdoptOpenJDK

covers1624 committed:
Update to Gradle 8.10

covers1624 committed:
Use settings.gralde to control repo list.

covers1624 committed:
Use version catalogs.

covers1624 committed:
Fix weirdness with wrapper sub-project.

covers1624 committed:
Start config parsing, stub out lex/neoforge modules.

covers1624 committed:
Switch to service loading gradle extract.

covers1624 committed:
Move Gradle & Extraction framework to sub-project.

covers1624 committed:
Refactor MCForge.

covers1624 committed:
Emit exceptions to the project logger before throwing.

Helps debug cryptic errors when running older Gradle versions.

covers1624 committed:
Add helper for getting PluginData from ProjectData.

covers1624 committed:
Move gradle tooling api to a version catalog.

covers1624 committed:
Add javadoc to Project/Plugin builders.

covers1624 committed:
Add ProjectTransformer concept to gradle extraction.

covers1624 committed:
Add gradle.properties helper to ProjectEmitter for tests.

covers1624 committed:
Only extract configuration tree for the compile/runtime classpath.

covers1624 committed:
Add MCForge data extraction basics.

covers1624 committed:
Exclude MCForge 'minecraft' configuration from the tree. Add tests.

covers1624 committed:
Refactor forge tests to clarify they are userdev tests.

covers1624 committed:
Add GitRepoManager for managing a clone.

covers1624 committed:
Add some debug toStrings to ConfigurationData.

covers1624 committed:
Simplify configuration extraction.

Simplifies configuration extraction to just extract the final view of the compile/runtime class paths of each source set.

Don't extract the configuration tree.
Don't copy the configuration, remove any dependencies from the tree in-place.

covers1624 committed:
Stub NeoForge gradle extract plugin.

covers1624 committed:
Ensure both compile and runtime classpaths are empty in FG tests.

covers1624 committed:
Add NeoForge userdev extraction tests.

covers1624 committed:
Add NeoDev extraction tests.

- Fix issues with Gradle Project's not actually being singletons.
- Add infrastructure for running test extractions from existing gradle projects.

covers1624 committed:
De-static WorkspaceToolEnvironment

covers1624 committed:
Refactor WorkspaceToolEnvironment to just Environment.

covers1624 committed:
Move Gradle plugin dev metadata to the Gradle plugin.

Allows for running the main application in-dev.

covers1624 committed:
Refactor RunConfig to RunConfigTemplate.

covers1624 committed:
Expand module references and build project paths.

covers1624 committed:
Start extracting gradle project info for modules.

covers1624 committed:
Make parent files when writing the extracted Gradle data.

covers1624 committed:
Fix logging path in main application.

covers1624 committed:
Implement large chunks of Intellij handling.

Unlike previous Workspace tool versions where we had 4 project models:
- One for the Gradle data exported from Gradle.
- One for the manipulatable model inside the main wstool app, plus its interfaces.
- One for the 'workspace' representation, mostly controlled by the workspace module.
- One within the workspace module prior to emitting it.

We instead, have exactly 2 project models.
- One for the gradle extraction data.
- One abstraction interface, that the main app can use, implemented by the workspace modules.

The intellij module implements these modules as it requires for emitting the project. That being, source sets are actually modules in Intellij's implementation of the api, however, this is not exposed back to the main application.

covers1624 committed:
Fix regression extracting project data from some NeoForge versions.

covers1624 committed:
Fix some unit tests, add Check workflow.

covers1624 committed:
Fix running tests from Gradle.

This probably needs to be less stupid, but I don't really know how..

covers1624 committed:
Add Java toolchain version extraction. Minor refactors.

covers1624 committed:
Rewrite how properties are handled in HashContainer.

We now receive a PropertyEntry object, similar to a how Hashes are done. This means no more magic constants, just pass the entry around and check/update it.

covers1624 committed:
Initial work of processing framework modules.

covers1624 committed:
Emit the Java version in the Intellij workspace.

covers1624 committed:
Ignore the .wstool folder from the IntelliJ project.

covers1624 committed:
Move Gradle cache into `gradle` folder.

covers1624 committed:
Persist the Gradle data on Modules.

Not happy with this, perhaps we'll end up with generic data storage on modules.

covers1624 committed:
Actually implement modules() on IJWorkspaceBuilder

covers1624 committed:
Fix dependencies getting duplicated on modules.

covers1624 committed:
Add hacky dependencies between main and test source sets.

This is a mega hack, we should remove this when we fix it Gradle side.

covers1624 committed:
Emit dependencies in the Intellij project!

- Emit shared libraries for all used by the project.
- Select the highest version of a depdendency and de-duplicate. (we should actually do this per-classpath, as thats what Gradle would do, but for now we just do this globally and hope nothing breaks. Thats was older WorkspaceTool has been doing successfully for years.)
- Hardlink all dependencies into the project cache directory, to mitigate Gradle cache cleanup issues.

Further work:
- More robust hardlinking.
- Extract more Gradle data for project exported api dependencies to avoid marking all as exported by default.

covers1624 committed:
Replace dependencies with cross-module deps.

covers1624 committed:
Add NeoForge cross-module dependencies to all discovered mods.

covers1624 committed:
Fix extraction of lwjgl dependencies from NeoForge.

covers1624 committed:
Invalidate the cached Gradle files when the classpath changes.

covers1624 committed:
Implement new integration testing framework.

covers1624 committed:
Ensure program output is stable.

covers1624 committed:
Cleanup tests and use better assertj apis.

covers1624 committed:
Fix other tests.

covers1624 committed:
Fix inter source set dependencies not being detected in all cases.

Specifically fixes the case where the outputs of the other source set are directly added to the first's compile/runtime classpath.

covers1624 committed:
Fix issues with inter-project dependencies.

Whilst processing the classpaths of one source set, if we removed a project dependency from its configurations, which were inherited to another source sets's classpaths, we break the other source set extraction.

covers1624 committed:
It seems PROJECT_DIR is not valid in all cases..

Apparently my testing here was insufficient. We can just do this in the test framework instead.

covers1624 commented:
Big progress milestone. 

It's now able to export a workspace which compiles, without any hax to the NeoForge project, this includes all tests! Already a massive improvement to pre-rewrite.

Next up, run configs.

covers1624 committed:
Refactors.

covers1624 committed:
Some more logging.

covers1624 committed:
Fix integration tests not visiting all files.

covers1624 committed:
Update to Java 21 for core wstool.

covers1624 committed:
Fix artifact classifiers being used as extensions.

covers1624 committed:
Support for extracting additional configurations from a Gradle Project.

covers1624 committed:
Fix extraction issues with newer NeoGradle versions.

covers1624 committed:
Add infrastructure for downloading Minecraft assets.

- Allows copying assets from existing launcher caches.

covers1624 committed:
Better interface for calling core WorkspaceTool functions from modules.

covers1624 committed:
Implement RunConfigs.

covers1624 committed:
Fix issues linking dependencies within sub-modules.

covers1624 committed:
Implement run configs for NeoForge 1.21.1

covers1624 committed:
Fix neoforge runs.

covers1624 committed:
Run the NeoForge setup task.

covers1624 committed:
Ensure we process sub-modules when assigning NeoForge dependencies.

covers1624 committed:
Rewrite the wrapper.

- Wrapper is now much cleaner.
- Provide jvm arg support to the launch manifest.
- Allow the launch manifest to declare a wrapper version range for compatible wrapper versions.
- Allow the wrapper manifest to declare multiple classpaths and replacement pattern them into the args.
- Require that the launch manifest is signed with my gpg key.

covers1624 committed:
Fix output instability from non-ordered collections.

covers1624 committed:
Remove dead NeoForge framework type.

covers1624 committed:
Add service storage to Environment.

covers1624 committed:
Cleanup AssetDownloader results.

covers1624 committed:
Split gradle task running from GradleModelExtractor/ModuleProcessor.

covers1624 committed:
Implement interface injection and access transformers with JST.

covers1624 committed:
Fix NeoForge setup always running.

covers1624 committed:
Implement a better content root merging algorithm.

covers1624 committed:
Ignore exceptions from cleaning up after tests.

covers1624 committed:
Update junit and switch to parameterized tests.

covers1624 committed:
Add some basic test coverage over Gradle 9, fix project deps.

covers1624 committed:
Include nested Gradle project files in cache key.

covers1624 committed:
Fix issues with clean builds.

covers1624 committed:
Fix Gradle classpath stability.

covers1624 committed:
Add gpg signing.

covers1624 committed:
Implement out-of-dev Gradle classpath property.

covers1624 commented:
The MVP of the rework is done.

The 'main' branch has now been created and fast-fowarded to this + released.
